Graplix v1

ho991217 opened the pull request:
# Update to Graplix V1

This PR includes updates to v1.

## 🔥 New features
- Create schema and identifier with OpenFGA DSL
- Automatic Type Generation
- Implicit identify with `id` property

## 📝 Checklist
- [x] validator
- [x] typegen
- [x] resolver integration with graplix

ho991217 committed:
feat: add validate function

tonyfromundefined commented on src/dsl/errors.ts:
```diff
@@ -0,0 +1,31 @@
+export abstract class SingleError extends Error {
+  constructor(public message: string) {
+    super(message);
+  }
+
+  toString() {
+    return this.message;
+  }
+}
+
+export class UnimplementedSingleError extends SingleError {
+  constructor(featureName: string) {
+    super(`${featureName} is not implemented`);
+    this.name = "UnimplementedError";
+  }
+}
+
+abstract class MultipleError<T = SingleError> extends Error {
+  constructor(public errors: Array<T>) {
+    super(
+      `${errors.length} error${errors.length > 1 ? "s" : ""} occurred:\n\t* ${errors.join("\n\t* ")}\n\n`,
+    );
+    this.errors = errors;
+  }
+
+  toString() {
+    return this.message;
+  }
+}
+
+export class UnimplementedError extends MultipleError<UnimplementedSingleError> {}
```
에러 네이밍이 조금 더 명시적이면 좋을거같아서 아래와 같이 수정해봤어요~

- Single, Multiple 개념이 헷갈려서 이렇게 정리하면 좋을거같아요.
  - Single: 생략
  - Multiple: "Multiple" prefix 붙이기
- `toString()` 구현이 들어간걸로 보이는데, 기본적으로 `Error` 객체에 구현되어있지 않나요? 따로 override하신 이유가 궁금해요.

```suggestion
enum ErrorCode {
  UNIMPLEMENTED = "UNIMPLEMENTED",
}

export class UnimplementedError extends Error {
  code = ErrorCode.UNIMPLEMENTED;
  
  constructor(featureName: string) {
    super(`${featureName} is not implemented`);
  }
}

export class MultipleError<T extends Error> extends Error {
  constructor(public errors: Array<T>) {
    super(
      `${errors.length} error${errors.length > 1 ? "s" : ""} occurred:\n\t* ${errors.join("\n\t* ")}\n\n`,
    );
  }

  toString() {
    return this.message;
  }
}

export class MultipleUnimplementedError extends MultipleError<UnimplementedError> {}
```

ho991217 replied:
> * `toString()` 구현이 들어간걸로 보이는데, 기본적으로 `Error` 객체에 구현되어있지 않나요? 따로 override하신 이유가 궁금해요.

fga/language 쪽 error 구현체를 가져오다보니 그런 것 같아요. 그 외에 제안주신 내용은 반영되었어요! [4fe1ce1](https://github.com/daangn/graplix/pull/9/commits/4fe1ce129d7a0801953cbfee868ccaee3c2897fc)

tonyfromundefined commented on src/dsl/fga.ts:
```diff
@@ -0,0 +1,8 @@
+import { transformer } from "@openfga/syntax-transformer";
+
+export function fga([data]: TemplateStringsArray): ReturnType<
```
```suggestion
export function parse([data]: TemplateStringsArray): ReturnType<
```

ho991217 replied:
```graphql```

ho991217 replied:
그래프큐엘 함수에서 착안했어요!
```ts
const query = graphql``;
```

아래 용도로 생각했을 때, `parse`라는 이름보다 맥락을 전달하는데에 효과적이라고 생각해요!
```ts
const scheme = fga`
  model
    schema 1.1

  type team
    relations
      define member: [user]
`;
```

tonyfromundefined commented on src/dsl/validate.spec.ts:
좋네요! ㅎㅎ 엣지케이스들은 차차 쌓아나가면 좋을거같아요.

tonyfromundefined commented on src/utils/exceptions.ts:
```diff
@@ -0,0 +1,9 @@
+export class ExceptionCollector {
+  constructor(private errors: UnimplementedSingleError[]) {}
+
+  raiseUnimplementedError(featureName: string) {
+    this.errors.push(new UnimplementedSingleError(featureName));
+  }
```
```suggestion
  captureUnimplementedError(featureName: string) {
    this.errors.push(new UnimplementedSingleError(featureName));
  }
```

ho991217 replied:
Reflected at: [80e3610](https://github.com/daangn/graplix/pull/9/commits/80e3610fdef0b9ceb7637167e23cfed77de42f05)

tonyfromundefined reviewed (commented):
수고하셨어요! 저는 아래와 같은 컨벤션을 사용하고 있어서 어떻게 생각하시는지 + 혹시 괜찮으시다면 반영이 가능할지 궁금해요.

"모든 export class, export function는 해당 class와 function 이름과 같은 파일 명을 갖는다."
- as-is: `errors.ts` -> to-be: `UnimplementedError.ts`, `MultipleError.ts`와 같은식으로 분리
- as-is: `exceptions.ts` -> to-be: `ExceptionCollector.ts`로 이름 변경

`dsl` 폴더 이름을 `language`라는 이름으로 바꾸면 더 좋을거같은데 어떠세요? dsl이라고 하는 폴더명이 맥락이 좀 있는거같아서 풀고 싶어서요!

ho991217 committed:
Reanme dsl to language

ho991217 committed:
Rename raiseUnimplementedError to captureUnimplementedError

ho991217 committed:
refactor: reorganize exports and update import paths

Restructure the codebase to consolidate language-related exports and update import paths for better maintainability and clarity. This includes moving errors to the language module and adjusting imports accordingly.

ho991217 committed:
refactor(language): restructure error handling and validation

Refactor error handling by introducing `UnimplementedError` and `MultipleError` classes to replace the previous `SingleError` and `UnimplementedSingleError`. Update validation logic to use the new error classes and rename `ExceptionCollector` to improve clarity. This change enhances maintainability and simplifies the error handling structure.

ho991217 committed:
feat: add validation to FGA model transformation

The `validate` function is now called after transforming the DSL to JSON object to ensure the model is valid before returning the AST. This improves the robustness of the FGA model processing.

ho991217 committed:
feat: add validation for mixed 'or' operator in user type relations

ho991217 committed:
chore: update types

ho991217 committed:
Implement parser (#10)

tonyfromundefined committed:
Implement playground (#17)

Co-authored-by: dylan.lee <ho991217@kakao.com>
Co-authored-by: dylan.lee <dylan.lee@daangn.com>

tonyfromundefined commented:
- import 문 정리
- packages/core -> packages/graplix로 폴더 이름 변경
- identify 안쪽으로 옮기기
- changeset 작성

ho991217 committed:
chore: refactor folder name from core to graplix

ho991217 committed:
refactor: organize imports

ho991217 committed:
refactor: organized yarn.lock

ho991217 committed:
BREAKING CHANGE: merge identify function into resolver

ho991217 committed:
docs: wrote changeset

ho991217 committed:
chore: update dependencies

tonyfromundefined reviewed (approved):
고생하셨어요! npm에 보여지는건 packages/graplix의 README.md일거라서, 요것도 예쁘게 채워놓으면 좋을거같아요.
